### PR TITLE
Enhanced Independent Coverage Reporting

### PR DESCRIPTION
**Pull request by IANTHEREAL**
## Why are these changes needed?

Reference https://github.com/microsoft/autogen/pull/857#issuecomment-1838719621 discussion on CI failure

Each test may require different dependencies, so they need to be run separately. And each job needs to upload the coverage report in the same way.

## Checks

- [x] I've included any doc changes needed for https://microsoft.github.io/autogen/. See https://microsoft.github.io/autogen/docs/Contribute#documentation to build and test documentation locally.
- [ ] I've added tests (if relevant) corresponding to the changes introduced in this PR.
- [ ] I've made sure all auto checks have passed.

### CODE REVIEW

**Comment by rickyloynd-microsoft** [comment]
@sonichi fyi

@IANTHEREAL Thanks for this! Is the PR in response to an existing issue or PR? Can you provide more details on the goals of the PR?

**Comment by IANTHEREAL** [comment]
> @sonichi fyi
> 
> @IANTHEREAL Thanks for this! Is the PR in response to an existing issue or PR? Can you provide more details on the goals of the PR?

It comes from the discussion of the PR https://github.com/microsoft/autogen/pull/857, and I have updated the PR description.

File: .github/workflows/contrib-tests.yml [hunk 1]
```diff
@@ -111,6 +124,18 @@ jobs:
...
         run: |
           pytest test/agentchat/contrib/test_gpt_assistant.py
+      - name: Coverage
+        if: matrix.python-version == '3.10'
+        run: |
+          pip install coverage>=5.3
+          coverage run -a -m pytest test/agentchat/contrib/test_gpt_assistant.py
+          coverage xml
+      - name: Upload coverage to Codecov
+        if: matrix.python-version == '3.10'
```
Comment (sonichi) [inline .github/workflows/contrib-tests.yml:134]:
```suggestion
      - name: Test GPTAssistantAgent
        if: matrix.python-version != '3.11'
        run: |
          pytest test/agentchat/contrib/test_gpt_assistant.py
      - name: Coverage
        if: matrix.python-version == '3.11'
        run: |
          pip install coverage>=5.3
          coverage run -a -m pytest test/agentchat/contrib/test_gpt_assistant.py
          coverage xml
      - name: Upload coverage to Codecov
        if: matrix.python-version == '3.11'
```

File: .github/workflows/contrib-tests.yml [hunk 2]
```diff
@@ -136,6 +161,18 @@ jobs:
...
         run: |
           pytest test/agentchat/contrib/test_teachable_agent.py
+      - name: Coverage
+        if: matrix.python-version == '3.10'
+        run: |
+          pip install coverage>=5.3
+          coverage run -a -m pytest test/agentchat/contrib/test_teachable_agent.py
+          coverage xml
+      - name: Upload coverage to Codecov
+        if: matrix.python-version == '3.10'
```
Comment (sonichi) [inline .github/workflows/contrib-tests.yml:171]:
```suggestion
      - name: Test TeachableAgent
        if: matrix.python-version != '3.9'
        run: |
          pytest test/agentchat/contrib/test_teachable_agent.py
      - name: Coverage
        if: matrix.python-version == '3.9'
        run: |
          pip install coverage>=5.3
          coverage run -a -m pytest test/agentchat/contrib/test_teachable_agent.py
          coverage xml
      - name: Upload coverage to Codecov
        if: matrix.python-version == '3.9'
```

**Review by sonichi — COMMENTED** [review]
Small suggestions to diversify the python versions and reduce redundant work.

**Comment by IANTHEREAL** [comment]
@sonichi PTAL again

**Review by sonichi — APPROVED** [review]

**Pull request merged**